src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,64 +218,202 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
+                lookup = definer.defineClassAsLookup(true);
+            }
+
+            MethodHandle constructor;
+            try {
+                constructor = lookup.findConstructor(lookup.lookupClass(), MT_void_Lookup_MethodHandle_MethodHandle)
+                        .asType(MT_Object_Lookup_MethodHandle_MethodHandle);
+            } catch (Throwable ex) {
+                throw uncaughtException(ex);
+            }
+            return new ProxyClassInfo(constructor, lookup);
+        }
+    };
+
+    private static final List<ClassDesc> DEFAULT_RETHROWS = List.of(desc(RuntimeException.class), desc(Error.class));
+    private static final ClassDesc CD_UndeclaredThrowableException = desc(UndeclaredThrowableException.class);
+    private static final ClassDesc CD_WrapperInstance = desc(WrapperInstance.class);
+    private static final ClassDesc CD_Empty = desc(Empty.class);
+    private static final MethodTypeDesc MTD_MethodHandle_Empty = MethodTypeDesc.of(CD_MethodHandle, CD_Empty);
+    private static final MethodTypeDesc MTD_Class_Empty = MethodTypeDesc.of(CD_Class, CD_Empty);
+    private static final MethodTypeDesc MTD_void_Throwable = MethodTypeDesc.of(CD_void, CD_Throwable);
+    private static final MethodType MT_void_Lookup_MethodHandle_MethodHandle = methodType(void.class, Lookup.class, MethodHandle.class, MethodHandle.class);
+    private static final MethodType MT_Object_Lookup_MethodHandle_MethodHandle = MT_void_Lookup_MethodHandle_MethodHandle.changeReturnType(Object.class);
+    private static final MethodTypeDesc MTD_void_Lookup_MethodHandle_MethodHandle = desc(MT_void_Lookup_MethodHandle_MethodHandle);
+    private static final MethodTypeDesc MTD_void_Lookup_Class = MethodTypeDesc.of(CD_void, CD_MethodHandles_Lookup, CD_Class);
+    private static final MethodTypeDesc MTD_MethodHandle_MethodType = MethodTypeDesc.of(CD_MethodHandle, CD_MethodType);
+    private static final String ORIGINAL_TARGET_NAME = "originalTarget";
+
+    /**
+     * Creates an implementation class file for a given interface. One implementation class is
+     * defined for each method handle, with the same bytes but different class data:
+     * [interfaceInfo, wrapperInstanceTarget, methodtype1, methodtype2, ...]
+     *
+     * @param ifaceDesc the given interface
+     * @param methodName the name of the single abstract method
+     * @param methods the information for implementation methods
+     * @return the bytes of the implementation classes
+     */
+    private static byte[] createTemplate(ClassDesc proxyDesc, ClassDesc ifaceDesc, String methodName, List<LocalMethodInfo> methods) {
+        return Classfile.build(proxyDesc, clb -> {
+            clb.withSuperclass(CD_Object);
+            clb.withFlags(ACC_FINAL | ACC_SYNTHETIC);
+            clb.withInterfaceSymbols(ifaceDesc, CD_WrapperInstance);
+
+            // individual handle fields
+            clb.withField(ORIGINAL_TARGET_NAME, CD_MethodHandle, ACC_PRIVATE | ACC_FINAL);
```

Comment:
Would a @Stable field help here? E.g if the returned functional interface instance is stored in a `static final` field, it should enable better performance?

Reply:
(actually, not sure - as the class is saved in a class value cache, so probably adding stable won't make any difference).

Reply:
The class is loaded as a hidden class, so all final fields are truly final (unlike classes but like records), so @Stable is not needed.

The class value cache only change performance if the cache was used in a hot path. And currently a class value cache is not considered as a cache by the JIT, i.e. the JIT does not know that if the key is a constant, the value is a constant.